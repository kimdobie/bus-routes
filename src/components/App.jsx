import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router';

import SelectRoute from './SelectRoute';
import SelectDirection from './SelectDirection';
import ListStops from './ListStops';

class App extends React.Component {
  constructor(props) {
    super(props);
    const { route, direction } = this.props;
    this.state = {
      route,
      direction
    };
    this.setRoute = this.setRoute.bind(this);
    this.setDirection = this.setDirection.bind(this);
    this.setHistory = this.setHistory.bind(this);
  }

  setRoute(route) {
    this.setState({ route, direction: '' }, () => this.setHistory());
  }

  setDirection(direction) {
    this.setState({ direction }, () => this.setHistory());
  }

  setHistory() {
    const { route, direction } = this.state;
    const { history } = this.props;
    if (route === '') {
      return history.push(`/`);
    }
    if (direction === '') {
      return history.push(`/route/${route}`);
    }
    return history.push(`/route/${route}/direction/${direction}`);
  }

  render() {
    const { route, direction } = this.state;
    return (
      <div>
        <SelectRoute route={route} setRoute={this.setRoute} />
        {route !== '' && (
          <SelectDirection
            direction={direction}
            setDirection={this.setDirection}
            route={route}
            key={`dir-${route}`}
          />
        )}
        {route !== '' && direction !== '' && (
          <ListStops
            direction={direction}
            route={route}
            key={`stop-${route}-${direction}`}
          />
        )}
      </div>
    );
  }
}

// export default App;
export default withRouter(App);

App.propTypes = {
  route: PropTypes.string,
  direction: PropTypes.string,
  history: PropTypes.shape.isRequired
};
App.defaultProps = {
  route: '',
  direction: ''
};
