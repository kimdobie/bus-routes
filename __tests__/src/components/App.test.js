import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App from '../../../src/components/App';

const axios = require('axios');

const fs = require('fs');
const path = require('path');

configure({ adapter: new Adapter() });

describe('App tests', () => {
  let historyMock = '';
  beforeEach(() => {
    historyMock = { push: jest.fn() };
  });

  const routes = fs.readFileSync(
    path.join(__dirname, '../../../__fixtures__/routes.json')
  );
  const directions = fs.readFileSync(
    path.join(__dirname, '../../../__fixtures__/directions.json')
  );
  const stops = fs.readFileSync(
    path.join(__dirname, '../../../__fixtures__/stops.json')
  );
  axios.setRoutes(JSON.parse(routes));
  axios.setDirections(JSON.parse(directions));
  axios.setStops(JSON.parse(stops));

  test('Snapshot without routes or directions', async () => {
    // eslint-disable-next-line react/jsx-filename-extension
    const app = await shallow(<App.WrappedComponent history={historyMock} />);
    expect(app).toMatchSnapshot();
  });
  test('Snapshot with routes', async () => {
    // TODO: This test produces an incorrect snapshot without SelectDirection
    const app = await shallow(
      <App.WrappedComponent history={historyMock} routes='123' />
    );
    expect(app).toMatchSnapshot();
  });
  test('Snapshot with routes and direction', async () => {
    // TODO: This test produces an incorrect snapshot without SelectDirection nor ListStops
    const app = await shallow(
      <App.WrappedComponent
        history={historyMock}
        routes='123'
        direction='456'
      />
    );
    expect(app).toMatchSnapshot();
  });

  test('Set route via props', async () => {
    const app = await shallow(
      <App.WrappedComponent history={historyMock} route='123' />
    );

    expect(app.state().route).toEqual('123');
    expect(app.state().direction).toEqual('');
    expect(app.find('SelectRoute')).toHaveLength(1);
    expect(app.find('SelectDirection')).toHaveLength(1);
    expect(app.find('ListStops')).toHaveLength(0);
  });

  test('Set direction and route via props', async () => {
    const app = await shallow(
      <App.WrappedComponent history={historyMock} route='123' direction='456' />
    );

    expect(app.state().route).toEqual('123');
    expect(app.state().direction).toEqual('456');
    expect(app.find('SelectRoute')).toHaveLength(1);
    expect(app.find('SelectDirection')).toHaveLength(1);
    expect(app.find('ListStops')).toHaveLength(1);
  });

  test('Set only directions via props', async () => {
    const app = await shallow(
      <App.WrappedComponent history={historyMock} direction='456' />
    );

    expect(app.state().route).toEqual('');
    expect(app.state().direction).toEqual('456');
    expect(app.find('SelectRoute')).toHaveLength(1);
    expect(app.find('SelectDirection')).toHaveLength(0);
    expect(app.find('ListStops')).toHaveLength(0);
  });

  test('Set route', async () => {
    const app = await shallow(<App.WrappedComponent history={historyMock} />);

    expect(app.state().route).toEqual('');
    expect(app.state().direction).toEqual('');
    expect(app.find('SelectRoute')).toHaveLength(1);
    expect(app.find('SelectDirection')).toHaveLength(0);
    expect(app.find('ListStops')).toHaveLength(0);

    app.instance().setRoute('123');
    expect(app.state().route).toEqual('123');
    expect(app.state().direction).toEqual('');
    expect(app.find('SelectRoute')).toHaveLength(1);
    expect(app.find('SelectDirection')).toHaveLength(1);
    expect(app.find('ListStops')).toHaveLength(0);
    expect(historyMock.push.mock.calls[0][0]).toEqual('/route/123');
  });

  test('Set directions', async () => {
    const app = await shallow(
      <App.WrappedComponent history={historyMock} route='123' />
    );

    expect(app.state().route).toEqual('123');
    expect(app.state().direction).toEqual('');
    expect(app.find('SelectRoute')).toHaveLength(1);
    expect(app.find('SelectDirection')).toHaveLength(1);
    expect(app.find('ListStops')).toHaveLength(0);

    app.instance().setDirection('456');
    expect(app.state().route).toEqual('123');
    expect(app.state().direction).toEqual('456');
    expect(app.find('SelectRoute')).toHaveLength(1);
    expect(app.find('SelectDirection')).toHaveLength(1);
    expect(app.find('ListStops')).toHaveLength(1);
    expect(historyMock.push.mock.calls[0][0]).toEqual(
      '/route/123/direction/456'
    );
  });

  test('Reset directions to ""', async () => {
    const app = await shallow(
      <App.WrappedComponent history={historyMock} route='123' direction='456' />
    );

    expect(app.state().route).toEqual('123');
    expect(app.state().direction).toEqual('456');

    app.instance().setDirection('');

    expect(app.state().route).toEqual('123');
    expect(app.state().direction).toEqual('');
    expect(app.find('SelectRoute')).toHaveLength(1);
    expect(app.find('SelectDirection')).toHaveLength(1);
    expect(app.find('ListStops')).toHaveLength(0);
    expect(historyMock.push.mock.calls[0][0]).toEqual('/route/123');
  });

  test('Reset route to ""', async () => {
    const app = await shallow(
      <App.WrappedComponent history={historyMock} route='123' direction='456' />
    );

    expect(app.state().route).toEqual('123');
    expect(app.state().direction).toEqual('456');

    app.instance().setRoute('');

    expect(app.state().route).toEqual('');
    expect(app.state().direction).toEqual('');
    expect(app.find('SelectRoute')).toHaveLength(1);
    expect(app.find('SelectDirection')).toHaveLength(0);
    expect(app.find('ListStops')).toHaveLength(0);
    expect(historyMock.push.mock.calls[0][0]).toEqual('/');
  });
});
