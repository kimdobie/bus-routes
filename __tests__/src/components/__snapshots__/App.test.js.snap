// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests Snapshot with routes 1`] = `
<div>
  <SelectRoute
    route=""
    setDirection={[Function]}
    setRoute={[Function]}
  />
</div>
`;

exports[`App tests Snapshot with routes and direction 1`] = `
<div>
  <SelectRoute
    route=""
    setDirection={[Function]}
    setRoute={[Function]}
  />
</div>
`;

exports[`App tests Snapshot without routes or directions 1`] = `
<div>
  <SelectRoute
    route=""
    setDirection={[Function]}
    setRoute={[Function]}
  />
</div>
`;
